// @c8y/ngx-components/esm2020/core/docs/quick-links-outlet/quick-links-outlet.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '../../common/common.module';
import { hookDrawer } from '../../drawer/drawer.service';
import { ProductExperienceModule } from '../../product-experience/product-experience.module';
import { QuickLinksOutletComponent } from './quick-links-outlet.component';
import * as i0 from "@angular/core";
export class QuickLinksOutletModule {
}
QuickLinksOutletModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: QuickLinksOutletModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
QuickLinksOutletModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.0.6", ngImport: i0, type: QuickLinksOutletModule, declarations: [QuickLinksOutletComponent], imports: [CommonModule, ProductExperienceModule] });
QuickLinksOutletModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: QuickLinksOutletModule, providers: [
        hookDrawer({
            component: QuickLinksOutletComponent,
            position: 'right',
            priority: 60,
            id: 'quickLinks'
        })
    ], imports: [CommonModule, ProductExperienceModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: QuickLinksOutletModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, ProductExperienceModule],
                    declarations: [QuickLinksOutletComponent],
                    providers: [
                        hookDrawer({
                            component: QuickLinksOutletComponent,
                            position: 'right',
                            priority: 60,
                            id: 'quickLinks'
                        })
                    ]
                }]
        }] });
