// @c8y/ngx-components/esm2020/upgrade/upgrade.module.mjs
import { bridgeServiceProvider } from './bridge.service';
import { RouterModule, hookNavigator, hookTab, hookActionBar, hookAction, hookBreadcrumb, hookDocs } from '@c8y/ngx-components';
import { Ng1ActionBarFactoryService } from './action-bar.factory';
import { Ng1ActionFactoryService } from './action.factory';
import { Ng1BreadcrumbFactoryService } from './breadcrumb.factory';
import { Ng1NodesFactoryService } from './nodes.factory';
import { Ng1TabsFactoryService } from './tabs.factory';
import { NgModule } from '@angular/core';
import { Ng1DocsFactoryService } from './docs.factory';
import { authBridgeServiceProvider } from './auth-bridge.service';
import { absoluteDateServiceProvider } from './absolute-date.service';
import { smartRulesServiceProvider } from './smart-rules.service';
import * as angular from 'angular';
import { setAngularJSGlobal } from '@angular/upgrade/static';
import * as i0 from "@angular/core";
setAngularJSGlobal(angular);
export class UpgradeModule {
}
UpgradeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: UpgradeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
UpgradeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.0.6", ngImport: i0, type: UpgradeModule, imports: [RouterModule] });
UpgradeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: UpgradeModule, providers: [
        absoluteDateServiceProvider,
        bridgeServiceProvider,
        authBridgeServiceProvider,
        smartRulesServiceProvider,
        hookNavigator(Ng1NodesFactoryService),
        hookTab(Ng1TabsFactoryService),
        hookActionBar(Ng1ActionBarFactoryService),
        hookAction(Ng1ActionFactoryService),
        hookBreadcrumb(Ng1BreadcrumbFactoryService),
        hookDocs(Ng1DocsFactoryService)
    ], imports: [RouterModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: UpgradeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [RouterModule],
                    exports: [],
                    providers: [
                        absoluteDateServiceProvider,
                        bridgeServiceProvider,
                        authBridgeServiceProvider,
                        smartRulesServiceProvider,
                        hookNavigator(Ng1NodesFactoryService),
                        hookTab(Ng1TabsFactoryService),
                        hookActionBar(Ng1ActionBarFactoryService),
                        hookAction(Ng1ActionFactoryService),
                        hookBreadcrumb(Ng1BreadcrumbFactoryService),
                        hookDocs(Ng1DocsFactoryService)
                    ]
                }]
        }] });
