// @c8y/ngx-components/fesm2015/c8y-ngx-components-widgets-implementations-three-d-rotation-box-model.mjs
const model = {
    metadata: {
        version: 4.5,
        type: 'Object',
        generator: 'Object3D.toJSON'
    },
    geometries: [
        {
            uuid: '008a44cb-1e9a-411b-a60b-24033a399485',
            type: 'BufferGeometry',
            data: {
                attributes: {
                    position: {
                        itemSize: 3,
                        type: 'Float32Array',
                        array: [
                            -1, -1, -1, -1, -1, 1, -1, 1, -1, -1, -1, 1, -1, 1, 1, -1, 1, -1, -1, 1, -1, -1, 1, 1,
                            1, 1, -1, -1, 1, 1, 1, 1, 1, 1, 1, -1, 1, 1, -1, 1, 1, 1, 1, -1, -1, 1, 1, 1, 1, -1,
                            1, 1, -1, -1, -1, -1, 1, -1, -1, -1, 1, -1, 1, -1, -1, -1, 1, -1, -1, 1, -1, 1, 1, -1,
                            -1, -1, -1, -1, 1, 1, -1, -1, -1, -1, -1, 1, -1, 1, 1, -1, 1, 1, 1, -1, 1, 1, 1, -1,
                            1, -1, 1, 1, -1, -1, 1, 1, -1, 1
                        ],
                        normalized: false
                    },
                    normal: {
                        itemSize: 3,
                        type: 'Float32Array',
                        array: [
                            -1, 0, 0, -1, 0, 0, -1, 0, 0, -1, 0, 0, -1, 0, 0, -1, 0, 0, 0, 0, -1, 0, 0, -1, 0, 0,
                            -1, 0, 0, -1, 0, 0, -1, 0, 0, -1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0,
                            0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, -1, 0, 0, -1, 0, 0, -1, 0,
                            0, -1, 0, 0, -1, 0, 0, -1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0
                        ],
                        normalized: false
                    },
                    color: {
                        itemSize: 3,
                        type: 'Float32Array',
                        array: [
                            1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
                            1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
                            1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
                            1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1
                        ],
                        normalized: false
                    }
                },
                boundingSphere: {
                    center: [0, 0, 0],
                    radius: 1.7320508075688772
                }
            }
        }
    ],
    materials: [
        {
            uuid: '12611d36-1d7a-4452-9611-cc8f8930ba98',
            type: 'MeshStandardMaterial',
            name: 'Material.001',
            color: 4539084,
            roughness: 0.5,
            metalness: 0.5,
            emissive: 0,
            envMapIntensity: 1,
            vertexColors: true,
            depthFunc: 3,
            depthTest: true,
            depthWrite: true,
            colorWrite: true,
            stencilWrite: false,
            stencilWriteMask: 255,
            stencilFunc: 519,
            stencilRef: 0,
            stencilFuncMask: 255,
            stencilFail: 7680,
            stencilZFail: 7680,
            stencilZPass: 7680
        }
    ],
    object: {
        uuid: '9a04d950-5931-4eb9-bbbd-2214e520f6df',
        type: 'Group',
        name: 'box.gltf',
        layers: 1,
        matrix: [1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1],
        children: [
            {
                uuid: 'e0d36a1d-88cd-4f0b-83f3-cdd467b3416b',
                type: 'Mesh',
                name: 'Cube',
                userData: {
                    name: 'Cube'
                },
                layers: 1,
                matrix: [
                    -3.5, 0, 0, 0, 0, -1.3322676295501878e-15, 6, 0, 0, 0.203663, 4.5222270372846654e-17, 0,
                    -0.0451503, 0.0332229, -0.03196, 1
                ],
                geometry: '008a44cb-1e9a-411b-a60b-24033a399485',
                material: '12611d36-1d7a-4452-9611-cc8f8930ba98'
            }
        ]
    }
};

/**
 * Generated bundle index. Do not edit.
 */

export { model };
