// @c8y/ngx-components/esm2020/core/dynamic-component/dynamic-component.module.mjs
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CollapseModule } from 'ngx-bootstrap/collapse';
import { DynamicComponentComponent } from './dynamic-component.component';
import { DynamicComponentService } from './dynamic-component.service';
import { I18nModule } from '../i18n/i18n.module';
import { DynamicComponentAlertsComponent } from './dynamic-component-alerts.component';
import { CommonModule } from '../common';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
export class DynamicComponentModule {
}
DynamicComponentModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: DynamicComponentModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DynamicComponentModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.0.6", ngImport: i0, type: DynamicComponentModule, declarations: [DynamicComponentComponent, DynamicComponentAlertsComponent], imports: [CommonModule, CollapseModule, i1.RouterModule, I18nModule], exports: [DynamicComponentComponent] });
DynamicComponentModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: DynamicComponentModule, providers: [DynamicComponentService], imports: [CommonModule, CollapseModule, RouterModule.forChild([]), I18nModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: DynamicComponentModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [DynamicComponentComponent, DynamicComponentAlertsComponent],
                    imports: [CommonModule, CollapseModule, RouterModule.forChild([]), I18nModule],
                    exports: [DynamicComponentComponent],
                    providers: [DynamicComponentService]
                }]
        }] });
