// @c8y/ngx-components/esm2020/files-repository/files-repository-upload.component.mjs
import { Component, Output } from '@angular/core';
import { EventEmitter } from '@angular/core';
import { InventoryBinaryService } from '@c8y/client';
import { AlertService, GainsightService, gettext } from '@c8y/ngx-components';
import { PRODUCT_EXPERIENCE } from './files-repository-product-experience.constants';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/client";
import * as i2 from "@c8y/ngx-components";
export class FilesRepositoryUploadComponent {
    constructor(inventoryBinaryService, alertService, gainsightService) {
        this.inventoryBinaryService = inventoryBinaryService;
        this.alertService = alertService;
        this.gainsightService = gainsightService;
        /** Emits an event when modal is closed. `uploaded` indicates if any files were uploaded. */
        this.onClose = new EventEmitter();
        this.isLoading = false;
        this.dropFilesHereMessage = gettext('Drop files here');
    }
    async upload() {
        if (this.filesToUpload === undefined) {
            this.alertService.warning(gettext('No files to upload.'));
        }
        else {
            this.isLoading = true;
            const promises = [];
            try {
                this.filesToUpload.forEach(file => {
                    promises.push(this.inventoryBinaryService.create(file));
                });
                await Promise.all(promises);
                this.alertService.success(gettext('File(s) uploaded.'));
                this.onClose.emit({ uploaded: true });
                this.gainsightService.triggerEvent(PRODUCT_EXPERIENCE.EVENTS.FILE_REPOSITORY_UPLOAD, {
                    component: PRODUCT_EXPERIENCE.COMPONENTS.FILE_REPOSITORY_UPLOAD_COMPONENT,
                    action: PRODUCT_EXPERIENCE.RESULTS.UPLOAD_ITEM_SUCCESS
                });
            }
            catch (e) {
                if (e) {
                    this.alertService.addServerFailure(e);
                    this.gainsightService.triggerEvent(PRODUCT_EXPERIENCE.EVENTS.FILE_REPOSITORY_UPLOAD, {
                        component: PRODUCT_EXPERIENCE.COMPONENTS.FILE_REPOSITORY_UPLOAD_COMPONENT,
                        action: PRODUCT_EXPERIENCE.RESULTS.UPLOAD_ITEM_FAILED
                    });
                }
            }
            finally {
                this.isLoading = false;
            }
        }
    }
    onFileDroppedEvent(event) {
        if (event && event.length > 0) {
            this.filesToUpload = event;
        }
    }
}
FilesRepositoryUploadComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: FilesRepositoryUploadComponent, deps: [{ token: i1.InventoryBinaryService }, { token: i2.AlertService }, { token: i2.GainsightService }], target: i0.ɵɵFactoryTarget.Component });
FilesRepositoryUploadComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.0.6", type: FilesRepositoryUploadComponent, selector: "c8y-files-repository-upload", outputs: { onClose: "onClose" }, ngImport: i0, template: "<div class=\"c8y-viewport-modal\">\n  <div class=\"modal-header dialog-header\">\n    <i c8yIcon=\"upload\"></i>\n    <h4 translate id=\"modal-title\">Upload files</h4>\n  </div>\n  <div class=\"modal-inner-scroll\">\n    <div class=\"modal-body\" id=\"modal-body\">\n      <c8y-file-picker-new\n        #filePicker\n        [uploadInProgress]=\"isLoading\"\n        (dropped)=\"onFileDroppedEvent($event)\"\n      ></c8y-file-picker-new>\n    </div>\n  </div>\n  <div class=\"card-footer separator d-flex j-c-center\">\n    <button\n      (click)=\"onClose.emit({ uploaded: false })\"\n      [disabled]=\"isLoading\"\n      class=\"btn btn-default\"\n      title=\"{{ 'Cancel' | translate }}\"\n      translate\n      type=\"button\"\n    >\n      Cancel\n    </button>\n    <button\n      (click)=\"upload()\"\n      [disabled]=\"!filePicker.valid || isLoading\"\n      class=\"btn btn-primary btn-form\"\n      title=\"{{ 'Upload' | translate }}\"\n      translate\n      type=\"button\"\n      data-cy=\"c8y-files-repository-upload--upload\"\n    >\n      Upload\n    </button>\n  </div>\n</div>\n", dependencies: [{ kind: "directive", type: i2.IconDirective, selector: "[c8yIcon]", inputs: ["c8yIcon"] }, { kind: "directive", type: i2.C8yTranslateDirective, selector: "[translate],[ngx-translate]" }, { kind: "component", type: i2.FilePickerNewComponent, selector: "c8y-file-picker-new", inputs: ["accept", "maxAllowedFiles", "minRequiredFiles", "disableValidators", "uploadInProgress"], outputs: ["dropped"] }, { kind: "pipe", type: i2.C8yTranslatePipe, name: "translate" }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: FilesRepositoryUploadComponent, decorators: [{
            type: Component,
            args: [{ selector: 'c8y-files-repository-upload', template: "<div class=\"c8y-viewport-modal\">\n  <div class=\"modal-header dialog-header\">\n    <i c8yIcon=\"upload\"></i>\n    <h4 translate id=\"modal-title\">Upload files</h4>\n  </div>\n  <div class=\"modal-inner-scroll\">\n    <div class=\"modal-body\" id=\"modal-body\">\n      <c8y-file-picker-new\n        #filePicker\n        [uploadInProgress]=\"isLoading\"\n        (dropped)=\"onFileDroppedEvent($event)\"\n      ></c8y-file-picker-new>\n    </div>\n  </div>\n  <div class=\"card-footer separator d-flex j-c-center\">\n    <button\n      (click)=\"onClose.emit({ uploaded: false })\"\n      [disabled]=\"isLoading\"\n      class=\"btn btn-default\"\n      title=\"{{ 'Cancel' | translate }}\"\n      translate\n      type=\"button\"\n    >\n      Cancel\n    </button>\n    <button\n      (click)=\"upload()\"\n      [disabled]=\"!filePicker.valid || isLoading\"\n      class=\"btn btn-primary btn-form\"\n      title=\"{{ 'Upload' | translate }}\"\n      translate\n      type=\"button\"\n      data-cy=\"c8y-files-repository-upload--upload\"\n    >\n      Upload\n    </button>\n  </div>\n</div>\n" }]
        }], ctorParameters: function () { return [{ type: i1.InventoryBinaryService }, { type: i2.AlertService }, { type: i2.GainsightService }]; }, propDecorators: { onClose: [{
                type: Output
            }] } });
