// @c8y/ngx-components/esm2020/device-grid/device-grid-extension.service.mjs
import { Injectable, InjectFlags, InjectionToken, Injector } from '@angular/core';
import { Router } from '@angular/router';
import { ExtensionPointWithoutStateForPlugins, fromTriggerOnce, hookGeneric, PluginsResolveService } from '@c8y/ngx-components';
import { flatten } from 'lodash-es';
import { shareReplay } from 'rxjs/operators';
import { DeviceGridService } from './device-grid.service';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
import * as i2 from "@angular/router";
import * as i3 from "./device-grid.service";
/**
 * A hook to use for Multi Provider extension.
 * @deprecated Consider using the `hookDeviceGridAction` function instead.
 */
export const HOOK_DEVICE_GRID_ACTION = new InjectionToken('DeviceGridActionHook');
/**
 * You can either provide a single `DeviceGridActionExtension` as parameter:
 * ```typescript
 *  hookDeviceGridAction(...)
 * ```
 *
 * Or an array to directly register multiple:
 * ```typescript
 *  hookDeviceGridAction([...])
 * ```
 *
 * Or you provide an Service that implements `ExtensionFactory<DeviceGridActionExtension>`
 * ```typescript
 *  export class MyDeviceGridActionFactory implements ExtensionFactory<DeviceGridActionExtension> {...}
 *  ...
 *  hookDeviceGridAction(MyDeviceGridActionFactory)
 * ```
 * A typed alternative to `HOOK_DEVICE_GRID_ACTION_DELETE`.
 * @param action The `DeviceGridDeleteExtension`'s or `ExtensionFactory` to be provided.
 * @returns An `Provider` to be provided in your module.
 */
export function hookDeviceGridAction(action, options) {
    return hookGeneric(action, HOOK_DEVICE_GRID_ACTION, options);
}
/**
 * A service which defines device grid action options.
 */
export class DeviceGridExtensionService extends ExtensionPointWithoutStateForPlugins {
    constructor(rootInjector, plugins, router, deviceGridService) {
        super(rootInjector, plugins);
        this.router = router;
        this.deviceGridService = deviceGridService;
        this.items$ = this.setupItemsObservable();
    }
    setupItemsObservable() {
        return fromTriggerOnce(this.router, this.refresh$, [
            () => flatten(this.injectors.map(injector => injector.get(HOOK_DEVICE_GRID_ACTION, [
                {
                    type: "DELETE" /* BuiltInActionType.Delete */,
                    deviceMatches: (_mo) => true,
                    onAction: (mo) => this.deviceGridService.delete(mo),
                    refreshAfterActionDone: true
                }
            ], InjectFlags.Self))),
            () => this.factories
        ]).pipe(shareReplay(1));
    }
}
DeviceGridExtensionService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: DeviceGridExtensionService, deps: [{ token: i0.Injector }, { token: i1.PluginsResolveService }, { token: i2.Router }, { token: i3.DeviceGridService }], target: i0.ɵɵFactoryTarget.Injectable });
DeviceGridExtensionService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: DeviceGridExtensionService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: DeviceGridExtensionService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i0.Injector }, { type: i1.PluginsResolveService }, { type: i2.Router }, { type: i3.DeviceGridService }]; } });
