// @c8y/ngx-components/esm2020/core/application/app-switcher.service.mjs
import { Injectable } from '@angular/core';
import { get, orderBy } from 'lodash-es';
import { combineLatest } from 'rxjs';
import { map, shareReplay, take } from 'rxjs/operators';
import { AppStateService } from '../common';
import * as i0 from "@angular/core";
import * as i1 from "../common";
export class AppSwitcherService {
    constructor(ui) {
        this.ui = ui;
        this.visibleApplicationTypes = ['HOSTED', 'EXTERNAL'];
        const { currentTenant } = this.ui;
        this.appsOfCurrentUser$ = this.ui.currentAppsOfUser.pipe(map(apps => this.filterVisible(apps)), shareReplay(1));
        this.oneCloudApps$ = this.appsOfCurrentUser$.pipe(map(apps => apps.filter(app => this.isCloudApp(app))), map(cloudApps => this.orderApps(cloudApps)), shareReplay(1));
        const nonCloudApps$ = this.appsOfCurrentUser$.pipe(map(apps => apps.filter(app => !this.isCloudApp(app))));
        this.apps$ = combineLatest([nonCloudApps$, currentTenant]).pipe(map(([apps, tenant]) => this.filterDuplicates(apps, tenant)), map(apps => this.orderApps(apps)), shareReplay(1));
        this.finishedLoading$ = combineLatest([this.apps$, this.oneCloudApps$]).pipe(map(() => true), take(1), shareReplay(1));
    }
    filterVisible(apps) {
        return apps.filter(app => this.visibleApplicationTypes.includes(app.type) &&
            !app.noAppSwitcher &&
            !this.isPackage(app) &&
            !get(app, 'manifest.noAppSwitcher'));
    }
    isPackage(app) {
        return !!app.manifest?.isPackage;
    }
    isCloudApp(app) {
        return !!app.cloud;
    }
    filterDuplicates(apps, tenant) {
        // Filter out apps that have duplicate contextpaths and are not own owned by the current tenant;
        const filterFn = app => !apps.some(otherApp => app !== otherApp &&
            app.contextPath &&
            app.contextPath === otherApp.contextPath &&
            app.owner.tenant.id !== tenant.name);
        return apps.filter(filterFn);
    }
    orderApps(apps) {
        return orderBy(apps, ({ name }) => name.toLowerCase());
    }
}
AppSwitcherService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: AppSwitcherService, deps: [{ token: i1.AppStateService }], target: i0.ɵɵFactoryTarget.Injectable });
AppSwitcherService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: AppSwitcherService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: AppSwitcherService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: function () { return [{ type: i1.AppStateService }]; } });
