// @c8y/ngx-components/esm2020/core/authentication/authentication.module.mjs
import { NgModule } from '@angular/core';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { CommonModule } from '../common/common.module';
import { PasswordCheckListComponent } from './password-check-list.component';
import { PasswordConfirm } from './password-confirm.directive';
import { PasswordStrengthCheckerService } from './password-strength-checker.service';
import { PasswordStrengthComponent } from './password-strength.component';
import { NewPasswordComponent } from './new-password.component';
import { PasswordConfirmModalComponent } from './password-confirm-modal.component';
import { ModalModule } from '../modal/modal.module';
import { FormsModule } from '../forms/forms.module';
import { TotpChallengeComponent } from './totp-challenge.component';
import { SmsChallengeComponent } from './sms-challenge.component';
import { TotpSetupComponent } from './totp-setup.component';
import { PasswordService } from './password.service';
import { QRCodeModule } from 'angularx-qrcode';
import { ProvidePhoneNumberComponent } from './provide-phone-number.component';
import { CurrentPasswordModalComponent } from './current-password-modal.component';
import * as i0 from "@angular/core";
/**
 * The angular module definition for new password.
 * @exports PasswordStrengthComponent
 * @exports PasswordCheckListComponent
 * @exports PasswordConfirm
 * @exports NewPasswordComponent
 * @exports PasswordConfirmModalComponent
 */
export class AuthenticationModule {
}
AuthenticationModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: AuthenticationModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
AuthenticationModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.0.6", ngImport: i0, type: AuthenticationModule, declarations: [TotpChallengeComponent,
        PasswordConfirm,
        PasswordStrengthComponent,
        PasswordCheckListComponent,
        NewPasswordComponent,
        PasswordConfirmModalComponent,
        CurrentPasswordModalComponent,
        SmsChallengeComponent,
        TotpSetupComponent,
        ProvidePhoneNumberComponent], imports: [FormsModule, TooltipModule, CommonModule, ModalModule, QRCodeModule], exports: [TotpChallengeComponent,
        PasswordStrengthComponent,
        PasswordCheckListComponent,
        PasswordConfirm,
        NewPasswordComponent,
        PasswordConfirmModalComponent,
        CurrentPasswordModalComponent,
        SmsChallengeComponent,
        TotpSetupComponent,
        ProvidePhoneNumberComponent] });
AuthenticationModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: AuthenticationModule, providers: [PasswordStrengthCheckerService, PasswordService], imports: [FormsModule, TooltipModule, CommonModule, ModalModule, QRCodeModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: AuthenticationModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        TotpChallengeComponent,
                        PasswordConfirm,
                        PasswordStrengthComponent,
                        PasswordCheckListComponent,
                        NewPasswordComponent,
                        PasswordConfirmModalComponent,
                        CurrentPasswordModalComponent,
                        SmsChallengeComponent,
                        TotpSetupComponent,
                        ProvidePhoneNumberComponent
                    ],
                    imports: [FormsModule, TooltipModule, CommonModule, ModalModule, QRCodeModule],
                    exports: [
                        TotpChallengeComponent,
                        PasswordStrengthComponent,
                        PasswordCheckListComponent,
                        PasswordConfirm,
                        NewPasswordComponent,
                        PasswordConfirmModalComponent,
                        CurrentPasswordModalComponent,
                        SmsChallengeComponent,
                        TotpSetupComponent,
                        ProvidePhoneNumberComponent
                    ],
                    entryComponents: [PasswordConfirmModalComponent],
                    providers: [PasswordStrengthCheckerService, PasswordService]
                }]
        }] });
