// @c8y/ngx-components/esm2020/core/common/stringify-object.pipe.mjs
import { Pipe } from '@angular/core';
import { isDate as _isDate } from 'lodash-es';
import { DatePipe } from './date.pipe';
import * as i0 from "@angular/core";
import * as i1 from "./date.pipe";
export class StringifyObjectPipe {
    constructor(c8yDatePipe) {
        this.c8yDatePipe = c8yDatePipe;
        this.isoDateTimeRegex = /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2}(?:\.\d*)?)(([+-](\d{2}):(\d{2})|Z)?)$/;
    }
    transform(value) {
        if (!value) {
            return '';
        }
        if (typeof value === 'object' && !_isDate(value)) {
            return JSON.stringify(value);
        }
        if (this.isDate(value)) {
            return this.c8yDatePipe.transform(value);
        }
        return value;
    }
    isDate(value) {
        if (!value)
            return false;
        if (_isDate(value))
            return !isNaN(value);
        if (typeof value === 'string')
            return this.isoDateTimeRegex.test(value);
        return false;
    }
}
StringifyObjectPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: StringifyObjectPipe, deps: [{ token: i1.DatePipe }], target: i0.ɵɵFactoryTarget.Pipe });
StringifyObjectPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "14.0.6", ngImport: i0, type: StringifyObjectPipe, name: "stringifyObject" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: StringifyObjectPipe, decorators: [{
            type: Pipe,
            args: [{ name: 'stringifyObject' }]
        }], ctorParameters: function () { return [{ type: i1.DatePipe }]; } });
