// @c8y/ngx-components/esm2020/core/stepper/stepper.module.mjs
import { CdkStep, CdkStepperModule, STEPPER_GLOBAL_OPTIONS } from '@angular/cdk/stepper';
import { NgModule } from '@angular/core';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { CommonModule } from '../common/common.module';
import { C8yStepperButtons } from './stepper-buttons.component';
import { C8yStepperIcon } from './stepper-icon.directive';
import { C8yStepperProgress } from './stepper-progress.component';
import { C8yStepper } from './stepper.component';
import { StepperService } from './stepper.service';
import { StepperOutletComponent } from './stepper-outlet.component';
import { FormsModule } from '@angular/forms';
import * as i0 from "@angular/core";
/**
 * The angular module definition for c8y-stepper (Wizard).
 * @exports C8yStepper
 * @exports C8yStepperIcon
 * @exports C8yStepperButtons
 * @exports CdkStep
 */
export class StepperModule {
}
StepperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: StepperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
StepperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.0.6", ngImport: i0, type: StepperModule, declarations: [C8yStepper,
        C8yStepperIcon,
        C8yStepperProgress,
        C8yStepperButtons,
        StepperOutletComponent], imports: [CommonModule, CdkStepperModule, TooltipModule, FormsModule], exports: [C8yStepper, CdkStep, C8yStepperIcon, C8yStepperButtons, StepperOutletComponent] });
StepperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: StepperModule, providers: [
        StepperService,
        {
            /**
             * This enables additional states, like: edit, done and error
             * These states are used to display custom icons, if needed.
             * Any change in here will most probably result in not showing custom icons anymore
             */
            provide: STEPPER_GLOBAL_OPTIONS,
            useValue: { displayDefaultIndicatorType: false }
        }
    ], imports: [CommonModule, CdkStepperModule, TooltipModule, FormsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: StepperModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, CdkStepperModule, TooltipModule, FormsModule],
                    declarations: [
                        C8yStepper,
                        C8yStepperIcon,
                        C8yStepperProgress,
                        C8yStepperButtons,
                        StepperOutletComponent
                    ],
                    exports: [C8yStepper, CdkStep, C8yStepperIcon, C8yStepperButtons, StepperOutletComponent],
                    providers: [
                        StepperService,
                        {
                            /**
                             * This enables additional states, like: edit, done and error
                             * These states are used to display custom icons, if needed.
                             * Any change in here will most probably result in not showing custom icons anymore
                             */
                            provide: STEPPER_GLOBAL_OPTIONS,
                            useValue: { displayDefaultIndicatorType: false }
                        }
                    ]
                }]
        }] });
