// @c8y/ngx-components/esm2020/core/setup/setup.service.mjs
import { Injectable } from '@angular/core';
import { of } from 'rxjs';
import { map, switchMap } from 'rxjs/operators';
import { OptionsService } from '../common/options.service';
import { Permissions } from '../common/permissions.service';
import { AppStateService } from '../common/ui-state.service';
import { Steppers } from '../stepper/stepper.model';
import { StepperService } from '../stepper/stepper.service';
import { NEEDED_ROLE_FOR_SETUP, SETUP_FINISHED_STEP_ID } from './setup.model';
import * as i0 from "@angular/core";
import * as i1 from "../stepper/stepper.service";
import * as i2 from "../common/permissions.service";
import * as i3 from "../common/options.service";
import * as i4 from "../common/ui-state.service";
export class SetupService {
    /**
     * @ignore only DI and initialization
     */
    constructor(stepperService, permissions, options, ui) {
        this.stepperService = stepperService;
        this.permissions = permissions;
        this.options = options;
        this.ui = ui;
        this.isSetupNeeded$ = this.ui.currentApplication.pipe(map(app => this.ui.isOwnerOfApplication(app)), switchMap(isOwner => {
            if (!isOwner) {
                return of([]);
            }
            return this.stepperService.getById$(Steppers.SETUP);
        }), map((steps) => this.getNotCompletedSetupSteps(steps)), map(notCompletedSetupSteps => this.options.forceSetup ||
            (this.options.isSetup && this.needsSetup(notCompletedSetupSteps))));
    }
    needsSetup(notCompletedSetupSteps) {
        notCompletedSetupSteps = this.excludeSetupFinishStep(notCompletedSetupSteps);
        if (this.hasPermission()) {
            return notCompletedSetupSteps.length > 0;
        }
        // Show setup screen as there are required steps but the user
        // has no permission. The Setup screen will show an error to
        // the user.
        return this.hasRequiredSteps(notCompletedSetupSteps);
    }
    excludeSetupFinishStep(notCompletedSetupSteps) {
        return notCompletedSetupSteps.filter(({ setupId }) => setupId !== SETUP_FINISHED_STEP_ID);
    }
    hasRequiredSteps(notCompletedSetupSteps) {
        return notCompletedSetupSteps.some(step => step.required);
    }
    hasPermission() {
        return this.permissions.hasRole(NEEDED_ROLE_FOR_SETUP);
    }
    getNotCompletedSetupSteps(steps) {
        if (!this.ui.currentApplication.value) {
            return steps;
        }
        return steps.filter(step => step.setupId &&
            !(this.ui.currentApplication.value.config?.setup || []).includes(step.setupId));
    }
}
SetupService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: SetupService, deps: [{ token: i1.StepperService }, { token: i2.Permissions }, { token: i3.OptionsService }, { token: i4.AppStateService }], target: i0.ɵɵFactoryTarget.Injectable });
SetupService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: SetupService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: SetupService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: i1.StepperService }, { type: i2.Permissions }, { type: i3.OptionsService }, { type: i4.AppStateService }]; } });
