// @c8y/ngx-components/esm2020/core/docs/docs.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DocsService } from './docs.service';
import { GuideHrefDirective } from './guide-href.directive';
import { GuideDocsComponent } from './guide-docs.component';
import { QuickLinksOutletModule } from './quick-links-outlet/quick-links-outlet.module';
import { HelpAndSupportOutletModule } from './help-and-support-outlet/help-and-support-outlet.module';
import { LegalNoticesOutletModule } from './legal-notices-outlet/legal-notices-outlet.module';
import * as i0 from "@angular/core";
export class DocsModule {
}
DocsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: DocsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DocsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.0.6", ngImport: i0, type: DocsModule, declarations: [GuideHrefDirective, GuideDocsComponent], imports: [CommonModule,
        QuickLinksOutletModule,
        HelpAndSupportOutletModule,
        LegalNoticesOutletModule], exports: [GuideHrefDirective, GuideDocsComponent] });
DocsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: DocsModule, providers: [DocsService], imports: [CommonModule,
        QuickLinksOutletModule,
        HelpAndSupportOutletModule,
        LegalNoticesOutletModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: DocsModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [GuideHrefDirective, GuideDocsComponent],
                    imports: [
                        CommonModule,
                        QuickLinksOutletModule,
                        HelpAndSupportOutletModule,
                        LegalNoticesOutletModule
                    ],
                    exports: [GuideHrefDirective, GuideDocsComponent],
                    providers: [DocsService]
                }]
        }] });
