// @c8y/ngx-components/esm2020/core/file-picker-new/file-picker-new.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule as C8yCommonModule } from '../common/common.module';
import { FilePickerNewComponent } from './file-picker-new.component';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { FormsModule as C8yFormsModule } from '../forms/forms.module';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import * as i0 from "@angular/core";
/**
 * The angular module definition for a drop-area.
 * @exports FilePickerComponent
 */
export class FilePickerNewModule {
}
FilePickerNewModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: FilePickerNewModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FilePickerNewModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.0.6", ngImport: i0, type: FilePickerNewModule, declarations: [FilePickerNewComponent], imports: [C8yFormsModule, CommonModule, FormsModule, C8yCommonModule, TooltipModule], exports: [FilePickerNewComponent] });
FilePickerNewModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: FilePickerNewModule, imports: [C8yFormsModule, CommonModule, FormsModule, C8yCommonModule, TooltipModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: FilePickerNewModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [C8yFormsModule, CommonModule, FormsModule, C8yCommonModule, TooltipModule],
                    exports: [FilePickerNewComponent],
                    declarations: [FilePickerNewComponent]
                }]
        }] });
