// @c8y/ngx-components/esm2020/core/dashboard/dashboard.module.mjs
import { DragDropModule } from '@angular/cdk/drag-drop';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { BsDatepickerModule } from 'ngx-bootstrap/datepicker';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { ActionBarModule } from '../action-bar/action-bar.module';
import { BreadcrumbModule } from '../breadcrumb/breadcrumb.module';
import { CommonModule } from '../common/common.module';
import { DocsModule } from '../docs/docs.module';
import { DynamicComponentModule } from '../dynamic-component/dynamic-component.module';
import { HeaderModule } from '../header/header.module';
import { DashboardChildActionComponent } from './dashboard-child-action.component';
import { DashboardChildTitleComponent } from './dashboard-child-title.component';
import { DashboardChildComponent } from './dashboard-child.component';
import { DashboardComponent } from './dashboard.component';
import { WidgetTimeContextComponent } from './widget-time-context.component';
import { WidgetsDashboardComponent } from './widgets-dashboard.component';
import { DynamicFormsModule } from '../dynamic-forms';
import { HelpModule } from '../help';
import * as i0 from "@angular/core";
import * as i1 from "ngx-bootstrap/dropdown";
export class DashboardModule {
}
DashboardModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: DashboardModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DashboardModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.0.6", ngImport: i0, type: DashboardModule, declarations: [DashboardChildComponent,
        DashboardComponent,
        DashboardChildTitleComponent,
        DashboardChildActionComponent,
        WidgetsDashboardComponent,
        WidgetTimeContextComponent], imports: [CommonModule,
        ActionBarModule,
        DynamicComponentModule,
        DragDropModule, i1.BsDropdownModule, TooltipModule,
        HeaderModule,
        DocsModule,
        BreadcrumbModule,
        FormsModule,
        BsDatepickerModule,
        DynamicFormsModule,
        HelpModule], exports: [DashboardComponent,
        DashboardChildComponent,
        DashboardChildTitleComponent,
        DashboardChildActionComponent,
        WidgetsDashboardComponent,
        WidgetTimeContextComponent] });
DashboardModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: DashboardModule, imports: [CommonModule,
        ActionBarModule,
        DynamicComponentModule,
        DragDropModule,
        BsDropdownModule.forRoot(),
        TooltipModule,
        HeaderModule,
        DocsModule,
        BreadcrumbModule,
        FormsModule,
        BsDatepickerModule,
        DynamicFormsModule,
        HelpModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.6", ngImport: i0, type: DashboardModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        DashboardChildComponent,
                        DashboardComponent,
                        DashboardChildTitleComponent,
                        DashboardChildActionComponent,
                        WidgetsDashboardComponent,
                        WidgetTimeContextComponent
                    ],
                    imports: [
                        CommonModule,
                        ActionBarModule,
                        DynamicComponentModule,
                        DragDropModule,
                        BsDropdownModule.forRoot(),
                        TooltipModule,
                        HeaderModule,
                        DocsModule,
                        BreadcrumbModule,
                        FormsModule,
                        BsDatepickerModule,
                        DynamicFormsModule,
                        HelpModule
                    ],
                    exports: [
                        DashboardComponent,
                        DashboardChildComponent,
                        DashboardChildTitleComponent,
                        DashboardChildActionComponent,
                        WidgetsDashboardComponent,
                        WidgetTimeContextComponent
                    ]
                }]
        }] });
